README.md
# MiscellanyNotes README
All miscellany notes.  
This is a readme.
#### Github Markdown
[https://guides.github.com/features/mastering-markdown/](https://guides.github.com/features/mastering-markdown/)

#### CSDN Libary
[http://lib.csdn.net/](http://lib.csdn.net/)


#### GitHub Repertories
* [android-dev-cn](https://github.com/android-cn/android-dev-cn)
* [计算机相关技术资料整理](https://github.com/EZLippi/practical-programming-books)
* [C 语言资源大全中文版](https://github.com/jobbole/awesome-c-cn)
* [Python 资源大全中文版](https://github.com/jobbole/awesome-python-cn)
* [C++ 资源大全中文版](https://github.com/jobbole/awesome-cpp-cn)
* [Android 资源大全中文版](https://github.com/jobbole/awesome-android-cn)
* [机器学习资源大全中文版](https://github.com/jobbole/awesome-machine-learning-cn)
* [PHP 资源大全中文版](https://github.com/jobbole/awesome-php-cn)
* [Android官方培训课程中文版](https://github.com/kesenhoo/android-training-course-in-chinese)


#### Collections
* [编程类开放书籍荟萃](https://linuxstory.org/free-chinese-programming-books/)


### 仓库分离  
#### AI&ML&CV&DM&DL&PR&NLP&IR分离转移至MachineLearningYXF  
[https://github.com/yxfcodeup/MachineLearningYXF.git](https://github.com/yxfcodeup/MachineLearningYXF.git)  

#### Linux分离转移至LinuxNotes  
[https://github.com/yxfcodeup/LinuxNotes.git](https://github.com/yxfcodeup/LinuxNotes.git)  


#### ProgrammingLanguage分离转移至ProgrammingNotes  
[https://github.com/yxfcodeup/ProgrammingNotes.git](https://github.com/yxfcodeup/ProgrammingNotes.git)

Notes/News.md
# NEWS
***
### Hardwares
#### FPGA：下一代机器人感知处理器
* FPGA：高性能、低能耗、可编程
* 感知计算在FPGA上的加速
* FPGA与ROS机器人操作系统的结合
[http://geek.csdn.net/news/detail/73399](http://geek.csdn.net/news/detail/73399)

#### 重磅 | IBM 创造出世界上首个人工相变神经元，可实现高速无监督学习
[http://mp.weixin.qq.com/s?__biz=MzA3MzI4MjgzMw==&mid=2650717770&idx=1&sn=9f33fc6a5a3cabd55f26bd31e871769d&scene=23&srcid=0804Vt6d7YyRp59csFPvkeIy#rd](http://mp.weixin.qq.com/s?__biz=MzA3MzI4MjgzMw==&mid=2650717770&idx=1&sn=9f33fc6a5a3cabd55f26bd31e871769d&scene=23&srcid=0804Vt6d7YyRp59csFPvkeIy#rd)

#### 基于Deep Learning的中文分词尝试
[http://mp.weixin.qq.com/s?__biz=MzA4OTk5OTQzMg==&mid=2449231335&idx=1&sn=d3ba98841e85b7cea0049cc43b3c16ca#rd](http://mp.weixin.qq.com/s?__biz=MzA4OTk5OTQzMg==&mid=2449231335&idx=1&sn=d3ba98841e85b7cea0049cc43b3c16ca#rd)

#### A Complete Guide on Getting Started with Deep Learning in Python  
[http://mp.weixin.qq.com/s?__biz=MzA4MTA5MjE5Mw==&mid=2651788207&idx=1&sn=815fcfc4c83ed20ea673380f956d4756&chksm=8461f8a1b31671b727780dd094507e96f9ef0150fe8c872847aa3c30834e2e082dfd354209bd&scene=0#wechat_redirect](http://mp.weixin.qq.com/s?__biz=MzA4MTA5MjE5Mw==&mid=2651788207&idx=1&sn=815fcfc4c83ed20ea673380f956d4756&chksm=8461f8a1b31671b727780dd094507e96f9ef0150fe8c872847aa3c30834e2e082dfd354209bd&scene=0#wechat_redirect)

__OLD__/ProgrammingLanguage/Python/Numpy/numpy-100.py
#1. import the numpy package under the name np
import numpy as np


#1. print the numpy version and the configuration
#print(np.__version__)
#np.show_config()

#2. create a null vector of size 10
z = np.zeros(10)

#3. how to get the documentation of the numpy add function from the command line?
#python -c "import numpy; numpy.info(numpy.add)"
#print(np.info(np.add))

#4. create a null vector of size 10 but the fifth value which is 1
z = np.zeros(10) 
z[4] = 1

#5. create a vector with values ranging from 10 to 49
z = np.arange(10,50)

#6. reverse a vector (first element becomes last)
z = np.arange(50)
z = z[::-1]

#7. create a 3x3 matrix with values ranging from 0 to 8
z = np.arange(9).reshape(3,3)

#8. find indices of non-zero elements from [1,2,0,0,4,0]
z = np.nonzero([1,2,0,0,4,0])

#9. create a 3x3 identity matrix
z = np.eye(3)

#10. create a 10x10 array with randomi values and find the minimum and maximum values
z = np.random.random((10,10))
zmin , zmax = z.min() , z.max()

#11. create a random vector of size 30 and find the mean value
z = np.random.random(30)
z = z.mean()

#12. create a 2d array with 1 on the border and 0 inside
z = np.ones((10,10))
z[1:-1,1:-1] = 0

#13. what is the result of the following expression
z = 0 * np.nan
z = np.nan==np.nan
z = np.inf > np.nan
z = np.nan - np.nan
z = 0.3==3*0.1

#14. create a 5x5 matrix with values 1,2,3,4 just below the diagonal
z = np.diag(1+np.arange(4) , k=-1)

#15. create a  8x8 matrix and fill it with a checkerboard pattern
z = np.zeros((8,8) , dtype=int)
z[1::2 , ::2] = 1
z[::2 , 1::2] = 1

#16. consider a (6,7,8) shape array,what is the index (x,y,z) of the 100th element?
z = np.unravel_index(100 , (6,7,8))
z = np.arange(336).reshape(6,7,8)

#17. create a checkboard 8x8 matrix using the tile function
z = np.tile(np.array([[0,1],[1,0]]) , (4,4))

#18. Normalize a 5x5 random matrix
z = np.random.random((5,5))
zmax , zmin = z.max() , z.min()
z = (z - zmin) / (zmax - zmin)

#19. Multiply a 5x3 matrix by a 3x2 matrix
z = np.dot(np.ones((5,3)) , np.ones((3,2)))

#20. Given a 1D array,negate all elements which are between 3 and 8,in place.
z = np.arange(11)
z[(3<z) & (z<=8)] *= -1

#21. Create a 5x5 matrix with row values ranging from 0 to 4
z = np.zeros((5,5))
z += np.arange(5)

#22. Consider a generator function that generates 10 intergers and use it to build an array
def generate() :
    for x in xrange(10) :
        yield x
z = np.fromiter(generate() , dtype=float , count=-1)

#23. Create a vector of size 10 with values ranging from 0 to 1,both excluded
z = np.linspace(0 , 1 , 12 , endpoint=True)[1:-1]

#24. Create a random vector of size 10 and sort it
z = np.random.random(10)
z.sort()

#25. How to sum a small array faster than np.sum?
z = np.arange(10)
z = np.add.reduce(z)

#26. Consider two random array A and B,check if they ary equal
a = np.random.randint(0 , 2 , 5)
b = np.random.randint(0 , 2 , 5)
equal = np.allclose(a , b)
z = equal

#27. Make an array immutable(read-only)
z = np.zeros(10)
z.flags.writeable = False
#z[0] = 1 #ValueError: assignment destination is read-only

#28. Consider a random 10x2 matrix representing cartesian coordinates,convert them to polar coordinates
z = np.random.random((10 , 2))
x , y = z[:,0] , z[:,1]
r = np.sqrt(x**2 + y**2)
t = np.arctan2(y , x)

#29. Create random vector of size 10 and replace the maximum value by 0
z = np.random.random(10)
z[z.argmax()] = 0

#30. Create a structured array with x and y coordinates covering the [0,1]x[0,1] area
z = np.zeros((5 , 5) , [("x" , float) , ("y" , float)])
z["x"] , z["y"] = np.meshgrid(np.linspace(0 , 1 , 5) , np.linspace(0 , 1 , 5))

#31. Given two arrays,x and y,construct the Cauchy matrix c(cij = 1/(xi-yj))
x = np.arange(8)
y = x + 0.5
c = 1.0 / np.subtract.outer(x , y)
z = np.linalg.det(c)

#32. Print the minimum and maximum representable value for each numpy scalar type
"""
for dtype in [np.int8 , np.int32 , np.int64] :
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)
for dtype in [np.float32 , np.float64] :
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
"""

#33. How to print all the values of an array?
np.set_printoptions(threshold=np.nan)
z = np.zeros((25 , 25))

#34. How to find the closest value (to a given scalar) in an array?
z = np.arange(100)
v = np.random.uniform(0 , 100)
index = (np.abs(z-v)).argmin()
z = z[index]
z = index


print(z)




__OLD__/ProgrammingLanguage/Python/爬虫/taobaoSpider/src/taobao.py
#_______________________taobao.py_____________________
# -*- coding: utf-8 -*-

#-----------------------------------------------------------------------------------
# Name          : 
# Version       : 1.0.0
# Author        : yxf
# Language      : Python 2.7    
# Start time    : 2016-04-27 22:39
# End time      :
# Function      : 
# Operation     :
#-----------------------------------------------------------------------------------


import os
import sys
import time
import datetime
import thread
import threading
import re
import getopt
import ConfigParser
import logging
import logging.config
import logging.handlers
import urllib2
from urlparse import urlparse
from bs4 import *


os.environ["NLS_LANG"] = "SIMPLIFIED CHINESE_CHINA.UTF8"
script_path = os.path.split(os.path.split(os.path.realpath(__file__))[0])[0]
print("---> Path of " + str(sys.argv[0].split("/")[-1]) + ": " + str(script_path))
logging.config.fileConfig(str(script_path) + "/config/logging.conf")
logger = logging.getLogger("root")
conf_file = str(script_path) + "/config/main.conf"
main_conf = ConfigParser.ConfigParser()
main_conf.read(conf_file)

time_stamp = datetime.datetime.now().strftime("%Y-%m-%d")
path = os.getcwd()
keywords = []
keyword = ""


#Receiving storage paths and keywords
if len(argv) > 2 :
    path = argv[1]
    keywords = argv[2:]



def mainRun() :
    return


def optMain() :
    #logger.info("---> Catch command line argument...")
    logger.info("-----------------taobao------------------------")
    mainRun()


if __name__ == "__main__" :
    logger.info("---> Start...")
    optMain()


__OLD__/ProgrammingLanguage/Python/爬虫/taobaoSpider/taobaoSpiderScheduler.sh
#!/bin/bash

source ~/.bashrc
#source /etc/profile
workdir=`pwd`
hdfsdir="hdfs://192.168.1.130:8020/user/bp"
xdrdir="hdfs://192.168.1.130:8020/user/xdr"
outdir=${workdir}/out
logfile="${workdir}/logs/scheduler.log"
ip="192.168.1.130"
symbol=0

log()
{
	echo `date '+%Y-%m-%d %H:%M:%S'` $*
	echo `date '+%Y-%m-%d %H:%M:%S'` $* >> $logfile
}
log "log begin ================================================"
cd ${workdir}

nowtime=`date '+%S'`
datatime=`date +%Y%m%d%H -d '-4 hours'`
deletetime=`date +%Y%m%d%H -d '-30 days'`
dirtime=`date +%Y%m%d -d '-4 hours'`
datatime=2015060418
dirtime=20150604
hdfsfiles="${xdrdir}/${dirtime}/"
log "datatime:${datatime}"
log "dirtime:${dirtime}"
log "hdfsfiles:${hdfsfiles}"
log "spark-submit ${workdir}/src/userTags.py -p ${hdfsfiles} -t ${datatime}"

#spark-submit ${workdir}/src/userLabel.py -p ${hdfsfiles} -t ${dirtime}
spark-submit ${workdir}/src/test_userLabel.py -p ${hdfsfiles} -t ${dirtime}
log "monitor complete! -p ${hdfsfiles} -t ${datatime}"
mkdir -p ${workdir}/dataResults/${dirtime}
hdfs dfs -get ${hdfsdir}/${dirtime}/part-00000 ${workdir}/dataResults/${dirtime}
hdfs dfs -rm -r ${hdfsdir}/${dirtime}



__OLD__/ProgrammingLanguage/Python/爬虫/zhihuSpider/src/getQuestionsScrapy.py
#_______________________getQuestions.py_____________________
# -*- coding: utf-8 -*-

#-----------------------------------------------------------------------------------
# Name          : getQuestions
# Version       : 1.0.0
# Author        : yxf
# Language      : Python 2.7    
# Start time    : 2016-04-22 16:15
# End time      :
# Function      : 
# Operation     :
#-----------------------------------------------------------------------------------


import os
import sys
import time
import datetime
import thread
import threading
import re
import ConfigParser
import getopt
import logging
import logging.config
import logging.handlers
import urllib
import urllib2
import codecs


os.environ["NLS_LANG"] = "SIMPLIFIED CHINESE_CHINA.UTF8"
script_path = os.path.split(os.path.split(os.path.realpath(__file__))[0])[0]
print("---> Path of " + str(sys.argv[0].split("/")[-1]) + ": " + str(script_path))
logging.config.fileConfig(str(script_path) + "/config/logging.conf")
logger = logging.getLogger("root")
conf_file = str(script_path) + "/config/main.conf"
main_conf = ConfigParser.ConfigParser()
main_conf.read(conf_file)


def reverseResult(page) :
    return "{0},{1},{2},{3}".format(int(page[0]) , page[1] , int(page[2]) , int(page[3]))


def getOneWebPage(url , num) :
    enable_proxy = True
    proxy_handler = urllib2.ProxyHandler({"http":"58.222.254.11:3128" , "http":"61.185.219.126:3128" , "http":"218.247.161.37:80"})
    null_proxy_handler = urllib2.ProxyHandler({})
    user_agent = "Mozilla/5.0 (Windows NT 5.2) AppleWebKit/534.30 (KHTML, like Gecko) Chrome/12.0.742.122 Safari/534.30"
    headers = {'User-Agent':user_agent}
    values = {'username':'cqc' , 'password':'x1x'}
    question_id = int(num)
    question_name = ""
    focus = 0
    answer = 0
    web_page_file = ""
    name_rex = "<h2 class=\"zm-item-title zm-editable-content\">(\s*)(.+)(\s*)</h2>"
    focus_rex = "</button>(\s*)(\d+)(\s*)人关注该问题"
    answer_rex = "<h3 data-num=\"(\d+)\" id=\"zh-question-answer-num\">(\d+)(\s*)个回答</h3>"
    request = urllib2.Request(url)
    try :
        response = urllib2.urlopen(request , timeout=600)
        web_page_file = response.read()
    except Exception , e :
        logger.error("--------------------Error--------------------")
        logger.error("---> Error url: " + url)
        logger.error("---> " + str(e))
        logger.error("--------------------Error--------------------")
    finally :
        a = 0
    if ""==web_page_file or None==web_page_file :
        return False
    is_exist_rex_1 = re.search(re.compile("你似乎来到了没有知识存在的荒原...") , web_page_file)
    is_exist_rex_2 = re.search(re.compile("来源链接是否正确？用户、话题或问题是否存在？") , web_page_file)
    if None!=is_exist_rex_1 and None!=is_exist_rex_2 :
        return False
    name_match = re.findall(re.compile(name_rex) , web_page_file)
    #logger.info("---> " + str(name_match))
    if 1 != len(name_match) :
        return False
    if len(name_match[0]) < 2 :
        return False
    question_name = name_match[0][1]
    #print(question_name)
    focus_match = re.findall(re.compile(focus_rex) , web_page_file)
    #logger.info("---> " + str(focus_match))
    if 1 != len(focus_match) :
        return False
    if len(focus_match[0]) < 2 :
        return False
    focus = str(focus_match[0][1])
    answer_match = re.findall(re.compile(answer_rex)  , web_page_file)
    #logger.info("---> " + str(answer_match))
    if 1 != len(answer_match) :
        return False
    if len(answer_match[0]) < 2 :
        return False
    answer = str(answer_match[0][1])
    #with open(script_path + "/dataResults/test.txt" , 'w') as f :
    #    f.write(web_page_file)
    return [question_id , question_name , focus , answer]


def getMoreWebPages(start_num , end_num , thread_id) :
    origin_url = "https://www.zhihu.com/question/"
    results = []
    for i in xrange(start_num , end_num) :
        url = origin_url + str(i)
        logger.info("---> Url: " + url)
        res = getOneWebPage(url , i)
        if False == res :
            continue
        results.append(res)
    with open(script_path + "/dataResults/thread" + str(thread_id) , 'w') as f :
        for r in results :
            r = reverseResult(r)
            f.write(r)
            f.write("\n")


def mainRun() :
    #getMoreWebPages(27000000 , 27850629 , 1)
    #getOneWebPage("https://www.zhihu.com/question/27850529" , 27850529)
    all_start = 27000000
    all_end = 28000000
    num_thread = 100
    threads = []
    start_list = []
    end_list = []
    single_numbers = (all_end - all_start) / num_thread
    for x in range(num_thread) :
        start = single_numbers * x + all_start
        end = single_numbers * (x + 1) + all_start
        start_list.append(start)
        end_list.append(end)
    logger.info("---> Start_list: " + str(start_list))
    logger.info("---> End_list: " + str(end_list))
    for i in range(num_thread) :
        t_name = "_thread_" + str(i)
        threads.append(threading.Thread(target=getMoreWebPages , name=t_name , args=(start_list[i] , end_list[i] , i)))
    logger.info("---> Threads: " + str(threads))
    for t in threads :
        t.setDaemon(True)
        t.start()
    for t in threads :
        t.join()


def optMain() :
    #logger.info("---> Catch command line argument...")
    logger.info("-----------------Get Questions from Zhihu------------------------")
    mainRun()


if __name__ == "__main__" :
    logger.info("---> Start...")
    optMain()


__OLD__/ProgrammingLanguage/Python/爬虫/sinaSpider/referenceMaterials/[源码] 爬取移动端微博信息 (强推)/2016-04-23/spider_selenium_sina_content.py
# coding=utf-8

"""  
Created on 2016-01-09 @author: Eastmount

功能: 爬取新浪微博用户的信息
信息：用户ID 用户名 注册时间 性别 地址(城市) 是否认证 用户标签(明星、搞笑等信息)
    个人资料完成度 粉丝数 关注数 微博数 粉丝ID列表 关注人ID列表 特别关注列表
网址：http://weibo.cn/ 数据量更小 相对http://weibo.com/

"""    

import time            
import re            
import os    
import sys  
import codecs  
import shutil
import urllib 
from selenium import webdriver        
from selenium.webdriver.common.keys import Keys        
import selenium.webdriver.support.ui as ui        
from selenium.webdriver.common.action_chains import ActionChains


'''
版本过低
pip install -U selenium
WebDriverException: Message: Can't load the profile.
Profile Dir: %s If you specified a log_file in the FirefoxBinary constructor,
check it for details.
'''

#先调用无界面浏览器PhantomJS或Firefox    
#driver = webdriver.PhantomJS(executable_path="G:\phantomjs-1.9.1-windows\phantomjs.exe")    
driver = webdriver.Firefox()
#options = webdriver.ChromeOptions()
#options.add_experimental_option("excludeSwitches", ["ignore-certificate-errors"])
#driver = webdriver.Chrome(chrome_options=options)
wait = ui.WebDriverWait(driver,10)


#全局变量 文件操作读写信息
inforead = codecs.open("SinaWeibo_List_best_1.txt", 'r', 'utf-8')
infofile = codecs.open("SinaWeibo_Info_best_1.txt", 'a', 'utf-8')


#********************************************************************************
#                  第一步: 登陆weibo.cn 获取新浪微博的cookie
#        该方法针对weibo.cn有效(明文形式传输数据) weibo.com见学弟设置POST和Header方法
#                LoginWeibo(username, password) 参数用户名 密码

#        https://www.zhihu.com/question/21451510
#        http://www.cnblogs.com/fnng/p/3606934.html
#                             验证码暂停时间手动输入
#********************************************************************************

def LoginWeibo(username, password):
    try:
        #**********************************************************************
        # 直接访问driver.get("http://weibo.cn/5824697471")会跳转到登陆页面 用户id
        #
        # 用户名<input name="mobile" size="30" value="" type="text"></input>
        # 密码 "password_4903" 中数字会变动,故采用绝对路径方法,否则不能定位到元素
        #
        # 勾选记住登录状态check默认是保留 故注释掉该代码 不保留Cookie 则'expiry'=None
        #**********************************************************************
        
        #输入用户名/密码登录
        print u'准备登陆Weibo.cn网站...'
        driver.get("http://login.weibo.cn/login/")
        elem_user = driver.find_element_by_name("mobile")
        elem_user.send_keys(username) #用户名
        elem_pwd = driver.find_element_by_xpath("/html/body/div[2]/form/div/input[2]")
        elem_pwd.send_keys(password)  #密码
        #elem_rem = driver.find_element_by_name("remember")
        #elem_rem.click()             #记住登录状态

        #重点: 暂停时间输入验证码
        #pause(millisenconds)
        time.sleep(20)
        
        elem_sub = driver.find_element_by_name("submit")
        elem_sub.click()              #点击登陆
        time.sleep(2)
        
        #获取Coockie 推荐 http://www.cnblogs.com/fnng/p/3269450.html
        print driver.current_url
        print driver.get_cookies()  #获得cookie信息 dict存储
        print u'输出Cookie键值对信息:'
        for cookie in driver.get_cookies(): 
            #print cookie
            for key in cookie:
                print key, cookie[key]
                    
        #driver.get_cookies()类型list 仅包含一个元素cookie类型dict
        print u'登陆成功...'
        
        
    except Exception,e:      
        print "Error: ",e
    finally:    
        print u'End LoginWeibo!\n\n'


#********************************************************************************
#                  第二步: 访问个人页面http://weibo.cn/5824697471并获取信息
#                                VisitPersonPage()
#        编码常见错误 UnicodeEncodeError: 'ascii' codec can't encode characters 
#********************************************************************************

def VisitPersonPage(user_id):

    try:
        global infofile
        print u'准备访问个人网站.....'
        #原创内容 http://weibo.cn/guangxianliuyan?filter=1&page=2
        driver.get("http://weibo.cn/" + user_id)

        #**************************************************************************
        # No.1 直接获取 用户昵称 微博数 关注数 粉丝数
        #      str_name.text是unicode编码类型
        #**************************************************************************

        #用户id
        print u'个人详细信息'
        print '**********************************************'
        print u'用户id: ' + user_id

        #昵称
        str_name = driver.find_element_by_xpath("//div[@class='ut']")
        str_t = str_name.text.split(" ")
        num_name = str_t[0]      #空格分隔 获取第一个值 "Eastmount 详细资料 设置 新手区"
        print u'昵称: ' + num_name 

        #微博数 除个人主页 它默认直接显示微博数 无超链接
        #Error:  'unicode' object is not callable
        #一般是把字符串当做函数使用了 str定义成字符串 而str()函数再次使用时报错
        str_wb = driver.find_element_by_xpath("//div[@class='tip2']")  
        pattern = r"\d+\.?\d*"   #正则提取"微博[0]" 但r"(\[.*?\])"总含[] 
        guid = re.findall(pattern, str_wb.text, re.S|re.M)
        print str_wb.text        #微博[294] 关注[351] 粉丝[294] 分组[1] @他的
        for value in guid:
            num_wb = int(value)
            break
        print u'微博数: ' + str(num_wb)

        #关注数
        str_gz = driver.find_element_by_xpath("//div[@class='tip2']/a[1]")
        guid = re.findall(pattern, str_gz.text, re.M)
        num_gz = int(guid[0])
        print u'关注数: ' + str(num_gz)

        #粉丝数
        str_fs = driver.find_element_by_xpath("//div[@class='tip2']/a[2]")
        guid = re.findall(pattern, str_fs.text, re.M)
        num_fs = int(guid[0])
        print u'粉丝数: ' + str(num_fs)
        

        #***************************************************************************
        # No.2 文件操作写入信息
        #***************************************************************************

        infofile.write('=====================================================================\r\n')
        infofile.write(u'用户: ' + user_id + '\r\n')
        infofile.write(u'昵称: ' + num_name + '\r\n')
        infofile.write(u'微博数: ' + str(num_wb) + '\r\n')
        infofile.write(u'关注数: ' + str(num_gz) + '\r\n')
        infofile.write(u'粉丝数: ' + str(num_fs) + '\r\n')
        infofile.write(u'微博内容: ' + '\r\n\r\n')
        
        

        #***************************************************************************
        # No.3 获取关注人列表
        # http://weibo.cn/guangxianliuyan?filter=0&page=1
        # 其中filter=0表示全部 =1表示原创
        #***************************************************************************

        print '\n'
        print u'获取微博内容信息'
        num = 1
        while num <= 5:
            url_wb = "http://weibo.cn/" + user_id + "?filter=0&page=" + str(num)
            print url_wb
            driver.get(url_wb)
            #info = driver.find_element_by_xpath("//div[@id='M_DiKNB0gSk']/")
            info = driver.find_elements_by_xpath("//div[@class='c']")
            for value in info:
                print value.text
                info = value.text

                #跳过最后一行数据为class=c
                #Error:  'NoneType' object has no attribute 'groups'
                if u'设置:皮肤.图片' not in info:
                    if info.startswith(u'转发'):
                        print u'转发微博'
                        infofile.write(u'转发微博\r\n')
                    else:
                        print u'原创微博'
                        infofile.write(u'原创微博\r\n')
                        
                    #获取最后一个点赞数 因为转发是后有个点赞数
                    str1 = info.split(u" 赞")[-1]
                    if str1: 
                        val1 = re.match(r'\[(.*?)\]', str1).groups()[0]
                        print u'点赞数: ' + val1
                        infofile.write(u'点赞数: ' + str(val1) + '\r\n')

                    str2 = info.split(u" 转发")[-1]
                    if str2: 
                        val2 = re.match(r'\[(.*?)\]', str2).groups()[0]
                        print u'转发数: ' + val2
                        infofile.write(u'转发数: ' + str(val2) + '\r\n')

                    str3 = info.split(u" 评论")[-1]
                    if str3:
                        val3 = re.match(r'\[(.*?)\]', str3).groups()[0]
                        print u'评论数: ' + val3
                        infofile.write(u'评论数: ' + str(val3) + '\r\n')

                    str4 = info.split(u" 收藏 ")[-1]
                    flag = str4.find(u"来自")
                    print u'时间: ' + str4[:flag]
                    infofile.write(u'时间: ' + str4[:flag] + '\r\n')

                    print u'微博内容:'
                    print info[:info.rindex(u" 赞")]  #后去最后一个赞位置
                    infofile.write(info[:info.rindex(u" 赞")] + '\r\n')
                    infofile.write('\r\n')
                    print '\n'
                else:
                    print u'跳过', info, '\n'
                    break
            else:
                print u'next page...\n'
                infofile.write('\r\n\r\n')
            num += 1
            print '\n\n'
        print '**********************************************'
            
        
        
    except Exception,e:      
        print "Error: ",e
    finally:    
        print u'VisitPersonPage!\n\n'
        print '**********************************************\n'
        

    
#*******************************************************************************
#                                程序入口 预先调用
#*******************************************************************************
    
if __name__ == '__main__':

    #定义变量
    username = '1520161****'             #输入你的用户名
    password = '*********'               #输入你的密码
    
    #用户id url+id访问个人
    #user_id = 'renzhiqiang'
    #user_id = 'guangxianliuyan'  

    

    #操作函数
    LoginWeibo(username, password)      #登陆微博

    #driver.add_cookie({'name':'name', 'value':'_T_WM'})
    #driver.add_cookie({'name':'value', 'value':'c86fbdcd26505c256a1504b9273df8ba'})

    #注意
    #因为sina微博增加了验证码,但是你用Firefox登陆一次输入验证码,再调用该程序即可,因为Cookies已经保证
    #会直接跳转到明星微博那部分,即: http://weibo.cn/guangxianliuyan
    

    #在if __name__ == '__main__':引用全局变量不需要定义 global inforead 省略即可
    print 'Read file:'
    user_id = inforead.readline()
    while user_id!="":
        user_id = user_id.rstrip('\r\n')
        VisitPersonPage(user_id)         #访问个人页面
        user_id = inforead.readline()
        #break
    
    infofile.close()
    inforead.close()
    
    

    


__OLD__/ProgrammingLanguage/Python/爬虫/zhihuSpider/src/getQuestions.py
#_______________________getQuestions.py_____________________
# -*- coding: utf-8 -*-

#-----------------------------------------------------------------------------------
# Name          : getQuestions
# Version       : 1.0.0
# Author        : yxf
# Language      : Python 2.7    
# Start time    : 2016-04-22 16:15
# End time      :
# Function      : 
# Operation     :
#-----------------------------------------------------------------------------------


import os
import sys
import time
import datetime
import thread
import threading
import re
import ConfigParser
import getopt
import logging
import logging.config
import logging.handlers
import urllib
import urllib2
import codecs


os.environ["NLS_LANG"] = "SIMPLIFIED CHINESE_CHINA.UTF8"
script_path = os.path.split(os.path.split(os.path.realpath(__file__))[0])[0]
print("---> Path of " + str(sys.argv[0].split("/")[-1]) + ": " + str(script_path))
logging.config.fileConfig(str(script_path) + "/config/logging.conf")
logger = logging.getLogger("root")
conf_file = str(script_path) + "/config/main.conf"
main_conf = ConfigParser.ConfigParser()
main_conf.read(conf_file)


def reverseResult(page) :
    return "{0},{1},{2},{3}".format(int(page[0]) , page[1] , int(page[2]) , int(page[3]))


def getOneWebPage(url , num) :
    enable_proxy = True
    proxy_handler = urllib2.ProxyHandler({"http":"58.222.254.11:3128" , "http":"61.185.219.126:3128" , "http":"218.247.161.37:80"})
    null_proxy_handler = urllib2.ProxyHandler({})
    user_agent = "Mozilla/5.0 (Windows NT 5.2) AppleWebKit/534.30 (KHTML, like Gecko) Chrome/12.0.742.122 Safari/534.30"
    headers = {'User-Agent':user_agent}
    values = {'username':'cqc' , 'password':'x1x'}
    question_id = int(num)
    question_name = ""
    focus = 0
    answer = 0
    web_page_file = ""
    name_rex = "<h2 class=\"zm-item-title zm-editable-content\">(\s*)(.+)(\s*)</h2>"
    focus_rex = "</button>(\s*)(\d+)(\s*)人关注该问题"
    answer_rex = "<h3 data-num=\"(\d+)\" id=\"zh-question-answer-num\">(\d+)(\s*)个回答</h3>"
    request = urllib2.Request(url)
    try :
        response = urllib2.urlopen(request , timeout=600)
        web_page_file = response.read()
    except Exception , e :
        #logger.error("--------------------Error--------------------")
        #logger.error("---> Error url: " + url)
        #logger.error("---> " + str(e))
        #logger.error("--------------------Error--------------------")
        a = 1
    finally :
        a = 0
    if ""==web_page_file or None==web_page_file :
        return False
    is_exist_rex_1 = re.search(re.compile("你似乎来到了没有知识存在的荒原...") , web_page_file)
    is_exist_rex_2 = re.search(re.compile("来源链接是否正确？用户、话题或问题是否存在？") , web_page_file)
    if None!=is_exist_rex_1 and None!=is_exist_rex_2 :
        return False
    name_match = re.findall(re.compile(name_rex) , web_page_file)
    #logger.info("---> " + str(name_match))
    if 1 != len(name_match) :
        return False
    if len(name_match[0]) < 2 :
        return False
    question_name = name_match[0][1]
    #print(question_name)
    focus_match = re.findall(re.compile(focus_rex) , web_page_file)
    #logger.info("---> " + str(focus_match))
    if 1 != len(focus_match) :
        return False
    if len(focus_match[0]) < 2 :
        return False
    focus = str(focus_match[0][1])
    answer_match = re.findall(re.compile(answer_rex)  , web_page_file)
    #logger.info("---> " + str(answer_match))
    if 1 != len(answer_match) :
        return False
    if len(answer_match[0]) < 2 :
        return False
    answer = str(answer_match[0][1])
    #with open(script_path + "/dataResults/test.txt" , 'w') as f :
    #    f.write(web_page_file)
    return [question_id , question_name , focus , answer]


def getMoreWebPages(start_num , end_num , thread_id) :
    origin_url = "https://www.zhihu.com/question/"
    results = []
    for i in xrange(start_num , end_num) :
        time.sleep(0.01)
        url = origin_url + str(i)
        #logger.info("---> Url: " + url)
        res = getOneWebPage(url , i)
        if False == res :
            continue
        results.append(res)
        if 0 == len(results)%1000 :
            with open(script_path + "/dataResults/thread" + str(thread_id) , 'w') as f :
                for r in results :
                    r = reverseResult(r)
                    f.write(r)
                    f.write("\n")
                results = []


def mainRun(all_start , all_end) :
    #getMoreWebPages(27000000 , 27850629 , 1)
    #getOneWebPage("https://www.zhihu.com/question/27850529" , 27850529)
    #all_start = 27000000
    #all_end = 28000000
    num_thread = 10
    threads = []
    start_list = []
    end_list = []
    single_numbers = (all_end - all_start) / num_thread
    for x in range(num_thread) :
        start = single_numbers * x + all_start
        end = single_numbers * (x + 1) + all_start
        start_list.append(start)
        end_list.append(end)
    logger.info("---> Start_list: " + str(start_list))
    logger.info("---> End_list: " + str(end_list))
    for i in range(num_thread) :
        t_name = "_thread_" + str(i)
        threads.append(threading.Thread(target=getMoreWebPages , name=t_name , args=(start_list[i] , end_list[i] , i)))
    logger.info("---> Threads: " + str(threads))
    for t in threads :
        t.setDaemon(True)
        t.start()
    for t in threads :
        t.join()


def optMain() :
    #logger.info("---> Catch command line argument...")
    logger.info("-----------------Get Questions from Zhihu------------------------")
    logger.info("---> Catch command line arguments...")
    if len(sys.argv) < 2 :
        logger.error("---> Program needs two arguments.\nExit...")
        sys.exit(1)
    opts , args = getopt.getopt(sys.argv[1:] , "s:e:")
    all_start = 0
    all_end = 0
    for opt,value in opts :
        if "-s" == opt :
            all_start = int(value)
        elif "-e" == opt :
            all_end = int(value)
        else :
            print(value)
    mainRun(all_start , all_end)


if __name__ == "__main__" :
    logger.info("---> Start...")
    optMain()
